src/commands/handlers/misc.js

```diff
@@ -100,52 +93,73 @@ const handlers = {
             hostname: params[3],
             server: params[4],
             real_name: realname,
-            away: is_away,
             num_hops_away: hops_away,
             channel: params[1],
-            channel_modes: chan_modes,
-            tags: command.tags
+            tags: command.tags,
+            unparsed_flags: unparsedFlags,
+            ...parsedFlags,
         });
     },
 
     RPL_WHOSPCRPL: function(command, handler) {
         const cache = handler.cache('who');
         if (!cache.members) {
             cache.members = [];
+            cache.type = 'whox';
         }
+
         const params = command.params;
 
-        // G = Gone, H = Here
-        const is_away = params[6][0].toUpperCase() === 'G';
+        if (cache.token === 0) {
+            // Token validation has already been attempted but failed,
+            // ignore this event as it will not be emitted
+            return;
+        }
+
+        if (!cache.token) {
+            const token = parseInt(params[1], 10) || 0;
+            if (token && handler.client.whox_token.validate(token)) {
+                // Token is valid, store it in the cache
+                cache.token = token;
+            } else {
+                // This event does not have a valid token so did not come from irc-fw,
+                // ignore it as the response order may differ
+                cache.token = 0;
+                return;
+            }
+        }
 
-        // get user channel modes
-        const net_prefixes = handler.network.options.PREFIX;
-        // filter PREFIX array against the prefix's in who reply returning matched PREFIX objects
-        const chan_prefixes = net_prefixes.filter(f => params[6].indexOf(f.symbol) > -1);
-        // use _.map to return an array of mode strings from matched PREFIX objects
-        const chan_modes = _.map(chan_prefixes, 'mode');
+        const { parsedFlags, unparsedFlags } = Helpers.parseWhoFlags(params[7], handler.network.options);
 
         // Some ircd's use n/a for no level, unify them all to 0 for no level
-        const op_level = !/^[0-9]+$/.test(params[9]) ? 0 : parseInt(params[9], 10);
+        const op_level = !/^[0-9]+$/.test(params[10]) ? 0 : parseInt(params[10], 10);
```

Comment:
Op levels aren't very well defined (and were implemented wrong in InspIRCd before 3.13.0) but according to the spec a lower op level means that the users rank is higher not lower.